components/sp-solutions/org.wso2.carbon.sp.solutions.twitter.analytics/widgets/PopularTweets/src/PopularTweets.jsx

```diff
@@ -14,62 +14,147 @@
         };
 
-        this.handleResize = this.handleResize.bind(this);
-        this.props.glContainer.on('resize', this.handleResize);
         this._handleDataReceived = this._handleDataReceived.bind(this);
-    }
+        this.showUnreadTweets = this.showUnreadTweets.bind(this);
 
-    handleResize() {
-        this.setState({width: this.props.glContainer.width, height: this.props.glContainer.height});
+        this.props.glContainer.on('resize', () =>
+            this.setState({
+                width: this.props.glContainer.width,
+                height: this.props.glContainer.height
+            })
+        );
     }
 
     componentDidMount() {
         super.getWidgetConfiguration(this.props.widgetID)
             .then((message) => {
-                super.getWidgetChannelManager().subscribeWidget(this.props.id, this._handleDataReceived, message.data.configs.providerConfig);
+                super.getWidgetChannelManager()
+                    .subscribeWidget(this.props.id, this._handleDataReceived, message.data.configs.providerConfig);
             })
     }
 
     componentWillUnmount() {
         super.getWidgetChannelManager().unsubscribeWidget(this.props.id);
     }
 
-    _handleDataReceived(setData) {
-        this.setState({
-            tweetData: setData.data,
-        });
+    _handleDataReceived(receivedData) {
+        let result = [];
+
+        if (this.state.tweetData.length === 0) {
+            receivedData.data.map(
+                tweet => {
+                    result.push({tweetID: tweet[0], id: tweet[1]})
+                }
+            );
+            //select 5 most recent tweets to display and the next 15(or less) tweets as unread tweets.
+            let numberOfRemovableTweets = result.length - MAX_DISPLAYED_TWEET_COUNT;
+            let removedTweets =
+                result.splice(MAX_DISPLAYED_TWEET_COUNT, numberOfRemovableTweets < 0 ? 0 : numberOfRemovableTweets);
+            numberOfRemovableTweets = removedTweets.length - MAX_UNREAD_TWEET_COUNT;
+            removedTweets.splice(MAX_UNREAD_TWEET_COUNT, numberOfRemovableTweets < 0 ? 0 : numberOfRemovableTweets);
+
+            this.setState({
+                tweetData: result,
+                unreadTweets: removedTweets
+            });
+        } else {
+            result = this.state.unreadTweets;
+            receivedData.data.map(
+                tweet => {
+                    result.push({tweetID: tweet[0], id: tweet[1]})
+                }
+            );
+            //sort by timestamp descending and select 15(or less) most recent tweets
+            result.sort(function (a, b) {
+                return b.id - a.id
+            });
+            let numberOfRemovableTweets = result.length - MAX_UNREAD_TWEET_COUNT;
+            result.splice(MAX_UNREAD_TWEET_COUNT, numberOfRemovableTweets < 0 ? 0 : numberOfRemovableTweets);
+
+            this.setState({
+                unreadTweets: result
+            });
+        }
+    }
+
+
```

Comment:
Shall we remove this additional lines?

Reply:
Fixed with https://github.com/wso2/analytics-solutions/pull/306